akka-http/server/src/test/scala/endpoints/akkahttp/server/EndpointsJsonSchemaTest.scala

```diff
@@ -86,10 +87,23 @@ class EndpointsJsonSchemaTest
         )
       }
 
+      // Valid URL and invalid entity (3)
+      Put("/user/42").withEntity(
+        ContentTypes.`text/plain(UTF-8)`,
+        "{\"name\":\"Alice\",\"age\":true}"
+      ) ~> testRoutes.updateUser ~> check {
+        handled shouldBe false
+        rejection shouldBe UnsupportedRequestContentTypeRejection(
```

Comment:
I’m not in favor of using Akka HTTP rejections, instead I would rather handle the request and return a 415 (or 400) response.

Reply:
Sorry for the long period of no response! I've been pondering what a good way forward for this sort of problem is. Although I understand your concern, returning directly a 400/415 would mean that routes produced by _endpoints_ would no longer interact well with other Akka HTTP routes.

Right now, we mingle routes from `endpoint.implementBy(...)` and routes written in Akka HTTP's DSL directly (to handle routes that don't have representations in _endpoints_). If the routes implemented by _endpoints_ start manually returning 400/415, operators like `route1 ~ route2` won't work the same anymore (`route2` won't be tried if `route1` bypasses Akka HTTP's rejection mechanism). Besides, this will make it virtually impossible to ensure that our 415 responses are uniform.

Reply:
I’ve described my reasoning for not using rejections in #352. I’m happy to reconsider my decision if you think this is not a viable approach.

> Right now, we mingle routes from `endpoint.implementBy(...)` and routes written in Akka HTTP's DSL directly (to handle routes that don't have representations in _endpoints_). If the routes implemented by _endpoints_ start manually returning 400/415, operators like `route1 ~ route2` won't work the same anymore (`route2` won't be tried if `route1` bypasses Akka HTTP's rejection mechanism).

The reason why routes implemented by _endpoints_ manually return 400/500 responses, is because that’s the only way I found to ensure clients, servers, and documentation are consistent together. If a route implemented by _endpoints_ returns a rejection, and then the rejection handler returns a response that _endpoints_ is unaware about, then clients would not be able to handle this response, and documentation may be inconsistent with the actual response.

> Besides, this will make it virtually impossible to ensure that our 415 responses are uniform.

I understand that this is maybe not the most desirable solution, but what would you think of reusing the responses returned by _endpoints_ to ensure that your Akka HTTP endpoints return responses that are uniform with the responses returned by _endpoints_? (this is what we do at work)

Reply:
Thanks for the link to the issue - I'm glad to see this is something you considered! Let me try to convince you this should change:

Having the right content type is like having the right method (GET/PUT/etc.): it is part of the definition of the endpoint (documented in the OpenAPI spec, specified by clients, used by servers). The wrong content type just means the endpoint shouldn't match.

> The reason why routes implemented by endpoints manually return 400/500 responses, is because that’s the only way I found to ensure clients, servers, and documentation are consistent together.

### Consistency between documentation and client/server

I have the following expectation from the documentation interpreters:

  1. tell me how to correctly call an endpoint (aka: right URI, parameters, JSON schema, content type, etc.)
  2. list special ways in which the endpoint may fail even if I did everything in 1. This should cover:
        - additional validation errors that the JSON schema may not cover (400)
        - business logic errors (500)

I think this is the best that _endpoints_ can really hope for. Trying to cover all the ways in which a web-server can fail is going to be undesirably noisy and depends on the web-server backend; I don't expect 414 or 431 to be documented in OpenAPI, even if they can probably be reliably triggered by pathological inputs on virtually any web-server.

I view a 415 sort of like a souped up 404 - none of the endpoints matched, but the server is helpfully trying to inform the user that one endpoint came close to matching. Consequently, it has no business being documented or handled by `endpoints`.

### Consistency between client and server

 I don't think this causes any issues around consistency between client/server. The clients will specify the right content type and the servers expect the right content type. No different than what is already being done for specifying the right method.

Reply:
> I don't think this causes any issues around consistency between client/server. The clients will specify the right content type and the servers expect the right content type. No different than what is already being done for specifying the right method.

Indeed, _endpoints_ client interpreters would not be impacted. The problem is _ad-hoc_ clients implemented based on the OpenAPI documentation. If the documentation does not mention that 415 is a possible response type, and what would be the type of the entity in such a case (JSON, text, ?), an _ad-hoc_ client might find the situation surprising.

But maybe the infrastructure required to properly document such response types would be too heavy compared to its benefits…? So far, we have something that is exhaustive: all possible response types appear in the documentation (including 400 and 500 errors), and this is useful because returning 400 or 500 errors is common in practice, so we want to properly document them.

Reply:
So, here are my thoughts: let’s start with the status code 415 (and not 400, as I was maybe suggesting before), and not include such 415 responses in the documentation (like you do in this PR). If this becomes an issue (people eventually need to have these responses documented in OpenAPI) we can change later.

I’m also OK to use rejection handlers (and `HttpErrorHandler` in Play interpreter) for now, we can change later to align with the way we handle 400 and 500 responses, if needed.

Reply:
Thanks for reporting back! Are there any outstanding changes needed for this PR then? I think everything is as you describe already:

  * status code 415 is being returned
  * 415 responses are provisionally omitted from OpenAPI documentation (and from the scaladocs too)
  * we provisionally use rejection handlers / `HttpErrorHandler` for these errors

Reply:
Indeed, I’ve had another look at the changes and I think this is good to go! Thanks for your work and your patience :)